Stats: update all-time component to use QuerySiteStats.

timmyc opened the pull request:
This branch updates the All Time component on the stats insights page to use `<QuerySiteStats />` and the global state tree - with the added benefit of turning All Time Stats into a pure component.

__To Test__
- Open up a stats insights page
- Validate the All Time component has the same data, and operates as it does in production ( loading state etc )

<img width="735" alt="stats_ _wordpress_com" src="https://cloud.githubusercontent.com/assets/22080/16031344/f44d6d9c-31ae-11e6-8534-aa1283ea07f8.png">


Test live: https://calypso.live/?branch=update/stats/redux-all-time

nylen commented on client/my-sites/stats/all-time/index.jsx:
```diff
@@ -78,3 +97,27 @@ export default React.createClass( {
+		views,
+		visitors,
+		views_best_day,
+		views_best_day_total
```
This looks a bit weird to me for a couple of reasons:

- Lots of props
- Naming convention of `views_best_day` is contrary to Calypso practices

Having lots of props is probably OK - this way each propType can be validated separately, but I think it would be good to fix the names here:

```js
return {
    ...
    visitors,
    viewsBestDay: views_best_day,
    viewsBestDayTotal: views_best_day_total
};
```

Alternatively, this cleanup could be done at a higher level, then you could use the spread operator to avoid some repetition.

```js
return {
   requesting: ...,
   query,
   siteId,
   ...allTimeData.stats
};
```

nylen replied:
Looking at #6049 which is very similar, I think my preferred solution would be to fix the names at the Redux selector level, then use the spread operator.  Though it could certainly be argued that we should leave the property names as-is for simplicity's sake.

nylen commented:
This tests out well for me.

I think `StatsAllTime` is now a good candidate for a pure function component - this would modernize the code and get rid of a bit of the repetition.

timmyc commented:
Thank you kindly for the review @nylen.  I took your feedback + some from @aduth that was left on #6135 - specifically:

- removed `React.createClass` and replaced with extending `Component`
- created a `Parsers` object in `state/stats/lists/utils.js` for logic that parses/transforms stats API data, and for this particular payload, `camelCase` the variables to make them calypso-ey.  + Tests for the new parser
- brought `classNames` in this component up to naming standards, and removed some legacy cruft like `module-content` and `is-non-en` logic

Also a rebase for good measure 🔑

nylen commented on client/state/stats/lists/utils.js:
```diff
@@ -14,3 +16,24 @@ import toPairs from 'lodash/toPairs';
+			parseData[ camelCase( key ) ] = value;
+		} );
+
+		return parseData;
```
`parsedData` instead of `parseData` perhaps?

nylen commented on client/state/stats/lists/utils.js:
```diff
@@ -14,3 +16,24 @@ import toPairs from 'lodash/toPairs';
 	return JSON.stringify( sortBy( toPairs( query ), ( pair ) => pair[ 0 ] ) );
 }
+
+export const Parsers = {
```
Based on our variable naming guidelines:

> Constructors intended for use with new should have a capital first letter (UpperCamelCase).

> Variables intended to be used as a constant can be defined with the SCREAMING_SNAKE_CASE naming convention. Note that while any variable declared using const could be considered a constant, in the context of our application this usage should usually be limited to top-level or exported module values.

Since this isn't a constructor intended for use with `new`, I think this should be either `PARSERS` (constant) or `parsers` (regular variable export).  Another option would be a separate file.

aduth replied:
>Based on our variable naming guidelines:

I'd be to blame for the suggestion. I agree it's not one that's been documented, but has been used in the past (maybe mostly by me 😄  [example](https://github.com/Automattic/wp-calypso/blob/master/client/lib/media/constants.js)). I guess it'd be hard to classify this usage (objects of constants?) so perhaps best to change to your alternatives.

nylen replied:
Perhaps a docs update is in order.  Thinking about it a bit more, I do remember seeing this pattern and I'm not sure I like either of my suggestions better.

nylen commented on client/state/stats/lists/selectors.js:
```diff
@@ -119,3 +120,28 @@ export function getSiteStatsPostsCountByDay( state, siteId, query, date ) {
+ * @param  {Object}  query    Stats query object
+ * @return {Array}            Parsed Data for the query
+ */
+export const getSiteStatsParsedData = createSelector(
```
From the `Parsed` name here, and the docblock comments, it isn't super clear what this does.  I'm thinking "normalize" may be a better verb than "parse" for this action.

nylen commented on client/state/stats/lists/selectors.js:
```diff
@@ -119,3 +120,28 @@ export function getSiteStatsPostsCountByDay( state, siteId, query, date ) {
+
+		if ( 'function' === typeof Parsers[ statType ] ) {
+			return Parsers[ statType ].call( this, data );
+		}
```
What are the other values that `statType` can take here, besides `'stats'` as in this PR?  Do you anticipate needing to perform different operations for different `statType`s?

nylen replied:
Per chat discussion, answer is yes - see [`lib/stats/stats-list/stats-parser.js`](https://github.com/Automattic/wp-calypso/blob/4544afe/client/lib/stats/stats-list/stats-parser.js)

timmyc replied:
Quite a few endpoint payloads are covered now in 
https://github.com/Automattic/wp-calypso/blob/master/client/lib/stats/stats-list/stats-parser.js - this particular one in the PR is currently not.  So moving forward we will likely see much of the same logic in the old stats parser repeated here, but in an ES6 fashion.

aduth commented on client/state/stats/lists/utils.js:
```diff
@@ -14,3 +16,24 @@ import toPairs from 'lodash/toPairs';
+		}
+
+		const parseData = {};
+		mapKeys( data.stats, ( value, key ) => {
```
The purpose of `mapKeys` is to reassign the property names of the object, so this can be simplified:

```js
return mapKeys( data.stats, ( value, key ) => camelCase( key ) );
```

https://lodash.com/docs#mapKeys

nylen commented:
This iteration also tests out well for me.  It was nice to see that old values from the persisted Redux state displayed immediately, then updated when the API request completed.

timmyc commented:
Chatted a bit with @nylen about this in slack and determined that that we should rename `Parsers` to `normalizers` to better align with post normalizer and other methods that perform transforms/additions based upon API responses.

Renamed the selector to match that in e1bbe6c along with some tests for that selector.

nylen commented:
Looks good & works well.  Could use a rebase before merging.
